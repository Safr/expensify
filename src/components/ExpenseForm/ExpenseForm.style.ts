import styled, { css } from 'styled-components';
import {
  blue,
  borderColor,
  darkRed, fontSizeLarge, fontSizeSmall, lSize, mSize, sSize, white } from '../../theme/variables';
import media from '../../theme/media';

export const InputGroup = styled.form`
  display: flex;
  flex-direction: column;
  max-width: 620px;
  margin: 0 auto;
  padding-left: 10px;
  padding-right: 10px;
  padding-top: 15px;
  margin-bottom: ${mSize};
  ${media.tablet`
    margin-bottom: ${lSize};
    padding-left: 20px;
    padding-right: 20px;
    padding-bottom: 0;
    max-width: 800px;
  `}
`;

export const TextInput = styled.input`
  height: 50px;
  margin-bottom: ${sSize};
  padding: ${sSize};
  font-size: ${fontSizeLarge};
  font-weight: 300;
  border: 1px solid #cacccd;
`;

export const TextArea = styled.textarea`
  height: 100px;
  margin-bottom: ${sSize};
  margin-top: ${sSize};
  padding: ${sSize};
  font-size: ${fontSizeLarge};
  font-weight: 300;
  border: 1px solid ${borderColor};
`;

export const Paragraph = styled.p`
  margin: 0 0 ${mSize} 0;
  font-style: italic;
`;

export const Button = styled.button`
  display: inline-block;
  padding: ${sSize};
  margin-top: ${mSize};
  margin-right: auto;
  font-size: ${fontSizeSmall};
  font-weight: 300;
  line-height: 1;
  color: ${white};
  background: ${blue};
  ${({ red }) => red && css`
    background: ${darkRed};
  `}
  border-radius: 3px;
  border: none;
  ${media.tablet`
    font-size: ${fontSizeLarge};
  `}
` as any;

