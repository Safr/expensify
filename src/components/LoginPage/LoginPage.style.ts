import styled from 'styled-components';
import { lSize, mSize } from '../../theme/variables';
import media from '../../theme/media';

export const LoginWrapper = styled.div`
  align-items: center;
  background: url('${require('./images/bg.jpg')}') no-repeat center bottom;
  background-size: cover;
  display: flex;
  height: 100vh;
  justify-content: center;
  width: 100vw;
`;

export const LoginInnerBox = styled.div`
  padding: ${mSize};
  background: #f2f4ee;
  border-radius: 3px;
  text-align: center;
  width: 30rem;
  ${media.tablet`
    width: 35rem;
    padding: ${mSize} ${lSize};
  `}
`;

export const Title = styled.h1`
  margin: 0 0 ${mSize} 0;
  line-height: 1;
`;
export const P = styled.p`
  margin: 3px;
`;
