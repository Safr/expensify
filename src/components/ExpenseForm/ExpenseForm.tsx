import React, { Component, ChangeEvent, FormEvent } from 'react';
import moment, { Moment } from 'moment';
import { SingleDatePicker } from 'react-dates';
import { ExpensesI } from '../../types';
import { Button, InputGroup, TextArea, TextInput, Paragraph } from './ExpenseForm.style';

interface Props {
  expense?: ExpensesI;
  onSubmit: (expense: object) => void;
}

interface State {
  description: string;
  note: string;
  amount: number | string;
  createdAt: Moment;
  calendarFocused: boolean;
  error: string;
}

class ExpenseForm extends Component<Props, State> {
  state = {
    description: this.props.expense ? this.props.expense.description : '',
    note: this.props.expense ? this.props.expense.note : '',
    amount: this.props.expense ? (this.props.expense.amount / 100).toString() : '',
    createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
    calendarFocused: false,
    error: '',
  };

  handleDescriptionChange = (e: ChangeEvent<HTMLInputElement>) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  }

  handleNoteChange = (e: ChangeEvent<HTMLTextAreaElement>) => {
    const note = e.target.value;
    this.setState(() => ({ note }));
  }

  handleAmountChange = (e: ChangeEvent<HTMLInputElement>) => {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  }

  handleDateChange = (createdAt: Moment) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  }

  handleFocusChange = ({ focused }: {focused: boolean}) => {
    this.setState(() => ({ calendarFocused: focused }));
  }

  handleSubmit = (e: FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: 'Please, provide description and amount' }));
    } else {
      this.setState(() => ({ error: '' }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note,
      });
    }
  }

  render() {
    return (
      <InputGroup onSubmit={this.handleSubmit}>
        {this.state.error && <Paragraph>{this.state.error}</Paragraph>}
        <TextInput
          type="text"
          placeholder="Description"
          value={this.state.description}
          onChange={this.handleDescriptionChange}
        />
        <TextInput
          type="text"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.handleAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          focused={this.state.calendarFocused}
          numberOfMonths={1}
          // tslint:disable-next-line:jsx-no-lambda
          isOutsideRange={() => false}
          onDateChange={this.handleDateChange}
          onFocusChange={this.handleFocusChange}
        />
        <TextArea
          placeholder="Add a note for your expense (optional)"
          value={this.state.note}
          onChange={this.handleNoteChange}
        />
        <Button>Save Expense</Button>
      </InputGroup>
    );
  }
}

export default ExpenseForm;
