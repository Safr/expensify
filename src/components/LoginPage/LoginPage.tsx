import React from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { startLogin } from '../../actions';
import { Button } from '../../components/ExpenseForm/ExpenseForm.style';
import { LoginWrapper, LoginInnerBox, Title, P } from './LoginPage.style';
import StoreState from '../../types';

interface Props {
  onStartLogin: () => Promise<object>;
  error: string;
}

export const LoginPage: React.SFC<Props> = ({ onStartLogin, error }) => (
  <LoginWrapper>
    <LoginInnerBox>
      <Title className="box-layout__title">Expensify App</Title>
      <P>{'It\'s time to get your expenses under control.'}</P>
      <Button red onClick={onStartLogin}>Login with Google</Button>
      {error && <P>{error}</P>}
    </LoginInnerBox>
  </LoginWrapper>
);

const mapStateToProps = (state: StoreState) => ({
  error: state.async.error,
});

const mapDispatchToProps = (dispatch: Dispatch<any>) => ({
  onStartLogin: () => dispatch(startLogin()),
});

export default connect(mapStateToProps, mapDispatchToProps)(LoginPage);
