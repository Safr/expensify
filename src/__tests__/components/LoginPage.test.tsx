import React from 'react';
import { shallow } from 'enzyme';
import { LoginPage } from '../../components/LoginPage/LoginPage';
import { Button } from '../../components/ExpenseForm/ExpenseForm.style';

describe('LoginPage Component', () => {
  test('should render correctly', () => {
    const wrapper = shallow(<LoginPage />);
    expect(wrapper).toMatchSnapshot();
  });
  test('should call startLogin on button click', () => {
    const onStartLogin = jest.fn();
    const wrapper = shallow(<LoginPage onStartLogin={onStartLogin} />);
    wrapper.find(Button).simulate('click');
    expect(onStartLogin).toHaveBeenCalled();
  });
});
