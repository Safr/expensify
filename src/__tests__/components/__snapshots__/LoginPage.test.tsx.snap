// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage Component should render correctly 1`] = `
<styled.div>
  <styled.div>
    <styled.h1
      className="box-layout__title"
    >
      Expensify App
    </styled.h1>
    <styled.p>
      It's time to get your expenses under control.
    </styled.p>
    <styled.button
      red={true}
    >
      Login with Google
    </styled.button>
  </styled.div>
</styled.div>
`;
